File: ICPC/OfficialICPC/21WF/C.cc
#include <cassert>
#include <cmath>
#include <iostream>
#include <numeric>

using LL = long long;

__int128 ksm(__int128 a, __int128 b) {
  __int128 res = 1;
  while (b) {
    if (b & 1) {
      res *= a;
    }
    a *= a;
    b >>= 1;
  }
  return res;
}

int main() {
  LL n, m;
  std::cin >> n >> m;
  for (LL v = 1; v <= 5000; ++v) {
    for (LL u = v - 1; u >= 1; --u) {
      if (std::gcd(u, v) != 1) {
        continue;
      }
      __int128 vv = ksm(v, n);
      __int128 uu = ksm(u, n);
      if (vv == -1 || uu == -1) {
        continue;
      }
      auto x = (__int128)m * (v - u) / (vv - uu);
      if ((__int128)m * (v - u) == x * (vv - uu) && x >= 1) {
        std::cout << (v - u) << " " << v << '\n';
        return 0;
      }
    }
  }
  std::cout << "impossible" << std::endl;
  return 0;
}

File: ICPC/OfficialICPC/21WF/C.py
import math


def ksm(a, b):
    res = 1
    while b:
        if b & 1:
            res *= a
            if res > 15625000000000000000000:
                return -1
        a = a**2
        if a > 15625000000000000000000:
            return -1
        b >>= 1
    return res


if __name__ == "__main__":
    n, m = map(int, input().split())
    for v in range(2, 5001):
        for u in range(v - 1, 0, -1):
            if math.gcd(u, v) != 1:
                continue
            vv = ksm(v, n)
            uu = ksm(u, n)
            if vv == -1 or uu == -1:
                continue
            if m * (v - u) % (vv - uu) == 0:
                assert math.gcd(v, v - u) == 1
                print(v - u, v)
                exit(0)
    print("impossible")
